### ISSUE TITLE
ApplyResponseCoreAsync vs ApplyResponseCore

### ISSUE THREAD

**Issue by BadriL** [aspnet/Security#86]
This is related to issue #76 but don't think this is same as that. `ApplyResponseCoreAsync` and `ApplyResponseCore` are named so that it gives an impression that one is async and the other is blocking version. That is true but then these two methods are called by the base class in response to different events. `ApplyResponseCore` is called when response headers are sent while `ApplyResponseCoreAsync` is called as part of tear down. First of all, the names of these methods do not reflect this fact. Second, I hope these two are not combined into one as part of issue #76. The requirement I have needs both the methods. I'm buffering the response (switching network stream to memory) on a case-by-case basis. When the response stream is not buffered, I need to run my code using `ApplyResponseCore`. On the other hand, if I buffer, `ApplyResponseCore` will not be called at all, until the time I write the content from memory back into the network stream. The code copying memory contents into network has to run during the response processing part of the middleware pipeline and hence I must use `ApplyResponseCoreAsync` for this reason. Now, when I do that `ApplyResponseCore` does get called as well, since headers get sent now. While `ApplyResponseCoreAsync` is guaranteed to run once, `ApplyResponseCore` is not. So, I'm setting a flag in the handler class to ensure the logic runs only once.

Long story short, the methods name do not convey what they do. It would be better to rename them and retain both the methods. Please do not eliminate one. Also, any help from the base class in ensuring `ApplyResponseCore` gets called once will be of great help.

**Comment by HaoK** [comment]
All async now, please reopen if there's any more issues once this PR https://github.com/aspnet/Security/pull/283 is in
